### ISSUE TITLE
Interior face attributes are not transfered from Mesh to ParMesh

### ISSUE THREAD

**Issue by tuckerbabcock** [mfem/mfem#3755]
Hi,

I'm trying to set interior face attributes in a serial `Mesh`, and the construct a ParMesh from it. If I iterate over the serial mesh's faces I can set the attributes (after adding a non-const overload of `GetFace` to mesh.hpp) as:
```
for (int i = 0; i < mesh->GetNumFaces(); i++)
{
   auto *face = mesh->GetFace(i);
   face->SetAttribute(10);
}
```
However, once I create a parallel mesh from the serial mesh with this constructor:
```
ParMesh(MPI_Comm comm, Mesh &mesh, int *partitioning_ = NULL, int part_method = 1);
```
these attributes are lost.

Is there a way for them to be retained when creating a parallel mesh?

Thanks,

Tucker

**Comment by v-dobrev** [comment]
Is your mesh a non-conforming mesh?

**Comment by tuckerbabcock** [comment]
Not non-conforming. I create and modify it just like this (again, after adding a non-const overload of `GetFace`):
```
// generate a simple tet mesh
std::unique_ptr<mfem::Mesh> mesh(
   new mfem::Mesh(mfem::Mesh::MakeCartesian2D(nxy, nxy,
                                  mfem::Element::TRIANGLE)));

// assign attributes to top and bottom sides
for (int i = 0; i < mesh->GetNE(); ++i)
{
   auto *elem = mesh->GetElement(i);

   mfem::Array<int> verts;
   elem->GetVertices(verts);

   bool below = true;
   for (int i = 0; i < verts.Size(); ++i)
   {
      auto *vtx = mesh->GetVertex(verts[i]);
      if (vtx[1] <= 0.5)
      {
         below = below;
      }
      else
      {
         below = false;
      }
   }
   if (below)
   {
      elem->SetAttribute(1);
   }
   else
   {
      elem->SetAttribute(2);
   }
}

// assign attributes to middle edge
for (int i = 0; i < mesh->GetNumFaces(); i++)
{
   auto *face = mesh->GetFace(i);

   mfem::Array<int> verts;
   face->GetVertices(verts);

   bool on_edge = true;
   for (int i = 0; i < verts.Size(); ++i)
   {
      auto *vtx = mesh->GetVertex(verts[i]);
      if (abs(vtx[1] - 0.5) < 1e-6)
      {
         on_edge = on_edge;
      }
      else
      {
         on_edge = false;
      }
   }
   if (on_edge)
   {
      face->SetAttribute(1);
   }
   else
   {
      face->SetAttribute(0);
   }
}

mesh->SetAttributes();

```

Then, I try to make a parallel mesh as
```
pmesh = std::make_unique<mfem::ParMesh>(comm, *mesh);
```

If I then iterate over the faces of the parallel mesh, the face attributes are all set to `1`, but the element attributes remain to what I'd set them to.

**Comment by v-dobrev** [comment]
Oh, you are setting attributes on (some) _face_ elements -- MFEM does not use these actually.

What you can do, instead, is to add "boundary" elements corresponding to the faces you want. The boundary attributes of these will then be propagated.

**Comment by tuckerbabcock** [comment]
Thanks for the reply @v-dobrev. Perhaps it would help if I explained a bit more what I'm really trying to accomplish. Similar to #3415, I'm trying to have two different domains coupled by a "special" model. In my case, I'm solving a temperature field problem with a thermal contact resistance interface between two materials.

Based on the suggestions in that issue, I'm using a DG space for my whole domain, and want a specific interior face integrator active only on the boundary between the two domains. I'm trying to determine when to use that integrator based on the attribute of the `FaceElementTransformations` passed to it.

If I add these internal boundaries, and then attribute them, will that affect anything else?
Or, can you suggest a more appropriate way to use/not use the interior face integrator?

Thanks,

Tucker

**Comment by tuckerbabcock** [comment]
@v-dobrev is there a way, when iterating over the interior faces as I am, to tell MFEM to turn the face into an internal boundary? Or to otherwise programmatically add internal boundary elements? I can't see anything in the `mesh.hpp` header.

Thanks,

Tucker

**Comment by hughcars** [comment]
Hi @tuckerbabcock, 

Based on your point above about these interfaces being between material domains, I think you could use this idiom:
```
   for (int f = 0; f < smesh.GetNumFaces(); ++f)
   {
      int e1, e2;
      smesh.GetFaceElements(f, &e1, &e2);
      if (e1 >= 0 && e2 >= 0 && smesh.GetAttribute(e1) != smesh.GetAttribute(e2))
      {
         // This is the internal face between attributes.
         auto *new_elem = smesh.GetFace(f)->Duplicate(&smesh);
         new_elem->SetAttribute(5);
         smesh.AddBdrElement(new_elem);
      }
   }
``` 
taken from [`hughcars/ncmesh-interior-boundary-dev`](https://github.com/mfem/mfem/tree/hughcars/ncmesh-interior-boundary-dev) in the file [`test_ncmesh.cpp`](https://github.com/mfem/mfem/blob/hughcars/ncmesh-interior-boundary-dev/tests/unit/mesh/test_ncmesh.cpp#L911-L924). This is adding a new boundary attribute `5`, on those faces that are discovered as being between different domain attributes. You could change this to check for particular values instead, but I was just looking for difference. 

Word of warning, if you're wanting to do AMR adaptivity with internal boundaries, as the PR that introduces that branch suggests, they're not currently supported (but will be very shortly hopefully).

**Comment by v-dobrev** [comment]
Just to add to @hughcars reply: after adding the new boundary elements, you should call the method `Mesh::SetAttributes` -- it updates the arrays `Mesh::attributes` and `Mesh::bdr_attributes` to be consistent with the current set of element and boundary attributes present in the `Mesh`.

**Comment by v-dobrev** [comment]
What integrators are you using and what method(s) do you use to add them to the `NonlinearForm`?

**Comment by lmdiazangulo** [comment]
We had this same issue for a custom integrator in DG. In this PR we tried to provide a solution: https://github.com/mfem/mfem/pull/3422

**Comment by v-dobrev** [comment]
Okay, I think I understand: you want to apply the DG integrator(s) just to the interior "boundary" faces. With the methods `AddInteriorFaceIntegrator` and `AddBdrFaceIntegrator` this is not possible -- the former applies the integrator to all interior faces, and the latter applies to just real boundary faces (i.e. faces that have just one neighboring volume element).

As @lmdiazangulo said, one needs to add a new type of method to handle this case. Take a look at his PR and maybe you can help review and finalize it.

**Comment by tuckerbabcock** [comment]
Thanks for the suggestions @lmdiazangulo and @v-dobrev. I've opened a PR (#3777) that implements a potential fix based on #3422. We can move the discussion there.
